File: src/systemd/query.rs
//! Service queries: list / info / status / file read / active check.

use std::collections::HashSet;
use std::process::Command;
use std::sync::OnceLock;
use std::time::{Duration, SystemTime};

use chrono::{DateTime, NaiveDateTime, TimeZone, Utc};
use regex::Regex;

use crate::error::{Error, Result};
use crate::systemd::manager::Manager;
use crate::systemd::mode::SystemdMode;
use crate::systemd::paths::{service_path, strip_prefix, systemd_unit_name};
use crate::systemd::types::ServiceInfo;

impl Manager {
    pub fn get_service_file(&self, name: &str) -> Result<String> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let p = service_path(name, self.mode)?;
        Ok(std::fs::read_to_string(p)?)
    }

    pub fn list_services(&self, all: bool) -> Result<Vec<ServiceInfo>> {
        let base = ["list-units", "--type=service", "--no-pager", "--no-legend"];
        let output = if all {
            self.exec.output(
                self.mode,
                &[base[0], base[1], base[2], base[3], "--all", "smdctl-*"],
            )
        } else {
            self.exec
                .output(self.mode, &[base[0], base[1], base[2], base[3], "smdctl-*"])
        };

        let output = match output {
            Ok(o) => o,
            Err(e) => {
                let es = e.to_string();
                if es.contains("No units found") {
                    return Ok(vec![]);
                }
                return Err(Error::msg(format!("list services: {es}")));
            }
        };

        let mut services = Vec::new();
        for line in output.lines() {
            let line = line.trim();
            if line.is_empty() {
                continue;
            }
            let fields: Vec<&str> = line.split_whitespace().collect();
            if fields.len() < 4 {
                continue;
            }
            let full_name = fields[0].trim_end_matches(".service");
            let name = strip_prefix(full_name);
            match self.get_service_info(&name) {
                Ok(info) => services.push(info),
                Err(_) => services.push(ServiceInfo {
                    name: name.clone(),
                    pid: 0,
                    status: fields[2].to_string(),
                    sub_state: fields.get(3).unwrap_or(&"").to_string(),
                    uptime: Duration::ZERO,
                    description: String::new(),
                    memory_bytes: 0,
                    ports: vec![],
                    mode: self.mode,
                }),
            }
        }
        Ok(services)
    }

    pub fn get_service_info(&self, name: &str) -> Result<ServiceInfo> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let unit = systemd_unit_name(name);

        let props = [
            "MainPID",
            "ActiveState",
            "SubState",
            "Description",
            "ActiveEnterTimestamp",
            "MemoryCurrent",
        ];

        let mut info = ServiceInfo {
            name: name.to_string(),
            pid: 0,
            status: String::new(),
            sub_state: String::new(),
            uptime: Duration::ZERO,
            description: String::new(),
            memory_bytes: 0,
            ports: vec![],
            mode: self.mode,
        };

        for prop in props {
            let Ok(value) = self.get_service_property(&unit, prop) else {
                continue;
            };
            match prop {
                "MainPID" => {
                    if let Ok(pid) = value.parse::<i32>() {
                        info.pid = pid;
                    }
                }
                "ActiveState" => info.status = value,
                "SubState" => info.sub_state = value,
                "Description" => info.description = value,
                "ActiveEnterTimestamp" => {
                    if !value.is_empty() && value != "n/a" {
                        if let Some(t) = parse_systemd_timestamp(&value) {
                            let now = SystemTime::now();
                            if let Ok(elapsed) = now.duration_since(t) {
                                info.uptime = elapsed;
                            }
                        }
                    }
                }
                "MemoryCurrent" => {
                    if !value.is_empty() && value != "[not set]" {
                        if let Ok(mem) = value.parse::<u64>() {
                            info.memory_bytes = mem;
                        }
                    }
                }
                _ => {}
            }
        }

        if matches!(self.mode, SystemdMode::User) && info.status == "active" && info.pid > 0 {
            info.ports = get_listening_ports(info.pid);
        }

        Ok(info)
    }

    fn get_service_property(&self, unit: &str, property: &str) -> Result<String> {
        let out = self
            .exec
            .output(self.mode, &["show", "-p", property, "--value", unit])?;
        Ok(out.trim().to_string())
    }

    pub fn get_status_text(&self, name: &str) -> Result<String> {
        if !self.service_exists(name) {
            return Err(Error::ServiceNotFound(name.to_string()));
        }
        let unit = systemd_unit_name(name);
        // `systemctl status` exits 3 when a unit is not running, so we accept
        // any output and treat exit code as informational.
        self.exec
            .output(self.mode, &["status", &unit, "--no-pager", "-l"])
            .or_else(|e| Ok(e.to_string()))
    }

    pub fn is_active(&self, name: &str) -> bool {
        let unit = systemd_unit_name(name);
        self.exec.status_ok(self.mode, &["is-active", &unit])
    }
}

fn parse_systemd_timestamp(ts: &str) -> Option<SystemTime> {
    // systemctl emits e.g. "Sat 2026-05-16 21:58:55 UTC" — the trailing "UTC" is
    // a literal, not an offset chrono can parse, so strip it and assume UTC.
    let trimmed = ts.trim().trim_end_matches(" UTC");
    if let Ok(naive) = NaiveDateTime::parse_from_str(trimmed, "%a %Y-%m-%d %H:%M:%S") {
        return Some(Utc.from_utc_datetime(&naive).into());
    }
    DateTime::parse_from_rfc3339(ts).ok().map(|d| d.into())
}

fn get_listening_ports(pid: i32) -> Vec<i32> {
    if pid <= 0 {
        return vec![];
    }
    let pid_str = pid.to_string();
    let mut ports = Vec::new();
    let mut seen = HashSet::new();

    if let Ok(tcp) = Command::new("ss").args(["-H", "-lntp"]).output() {
        append_ports_for_pid(
            &mut ports,
            &mut seen,
            &String::from_utf8_lossy(&tcp.stdout),
            &pid_str,
        );
    }
    if let Ok(udp) = Command::new("ss").args(["-H", "-lnup"]).output() {
        append_ports_for_pid(
            &mut ports,
            &mut seen,
            &String::from_utf8_lossy(&udp.stdout),
            &pid_str,
        );
    }
    ports.sort_unstable();
    ports
}

fn append_ports_for_pid(
    ports: &mut Vec<i32>,
    seen: &mut HashSet<i32>,
    output: &str,
    pid_str: &str,
) {
    static PID_RE: OnceLock<Regex> = OnceLock::new();
    let re = PID_RE.get_or_init(|| Regex::new(r"pid=(\d+)").unwrap());
    for line in output.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let owner = re
            .captures_iter(line)
            .any(|c| c.get(1).map(|m| m.as_str()) == Some(pid_str));
        if !owner {
            continue;
        }
        let fields: Vec<&str> = line.split_whitespace().collect();
        if fields.len() < 4 {
            continue;
        }
        let port = extract_port(fields[3]);
        if port > 0 && seen.insert(port) {
            ports.push(port);
        }
    }
}

fn extract_port(addr: &str) -> i32 {
    if let Some(idx) = addr.rfind("]:") {
        return addr[idx + 2..].parse().unwrap_or(0);
    }
    if let Some(idx) = addr.rfind(':') {
        return addr[idx + 1..].parse().unwrap_or(0);
    }
    0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_systemd_timestamp_utc_literal() {
        let parsed = parse_systemd_timestamp("Sat 2026-05-16 21:58:55 UTC")
            .expect("systemctl UTC timestamp must parse");
        let expected = Utc
            .with_ymd_and_hms(2026, 5, 16, 21, 58, 55)
            .single()
            .map(SystemTime::from)
            .unwrap();
        assert_eq!(parsed, expected);
    }

    #[test]
    fn parse_systemd_timestamp_rejects_n_a() {
        assert!(parse_systemd_timestamp("n/a").is_none());
    }
}

File: internal/systemd/query.go
package systemd

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

func (m *Manager) ListServices(all bool) ([]ServiceInfo, error) {
	args := []string{"list-units", "--type=service", "--no-pager", "--no-legend"}
	if all {
		args = append(args, "--all")
	}
	args = append(args, "smdctl-*")
	out, err := m.exec().Output(m.Mode, args...)
	if err != nil {
		if strings.Contains(err.Error(), "No units found") {
			return []ServiceInfo{}, nil
		}
		return nil, fmt.Errorf("list services: %w", err)
	}
	result := make([]ServiceInfo, 0)
	for _, line := range strings.Split(out, "\n") {
		f := strings.Fields(strings.TrimSpace(line))
		if len(f) < 4 {
			continue
		}
		name := StripPrefix(strings.TrimSuffix(f[0], ".service"))
		info, e := m.GetServiceInfo(name)
		if e != nil {
			info = ServiceInfo{Name: name, Status: f[2], SubState: f[3], Mode: m.Mode, Ports: []int{}}
		}
		result = append(result, info)
	}
	return result, nil
}
func (m *Manager) GetServiceInfo(name string) (ServiceInfo, error) {
	if !m.ServiceExists(name) {
		return ServiceInfo{}, fmt.Errorf("service not found: %s", name)
	}
	info := ServiceInfo{Name: name, Mode: m.Mode, Ports: []int{}}
	unit := UnitName(name)
	props := []string{"MainPID", "ActiveState", "SubState", "Description", "ActiveEnterTimestamp", "MemoryCurrent"}
	for _, p := range props {
		v, e := m.exec().Output(m.Mode, "show", "-p", p, "--value", unit)
		if e != nil {
			continue
		}
		v = strings.TrimSpace(v)
		switch p {
		case "MainPID":
			info.PID, _ = strconv.Atoi(v)
		case "ActiveState":
			info.Status = v
		case "SubState":
			info.SubState = v
		case "Description":
			info.Description = v
		case "ActiveEnterTimestamp":
			if t, ok := parseSystemdTimestamp(v); ok {
				if d := time.Since(t); d > 0 {
					info.Uptime = d
				}
			}
		case "MemoryCurrent":
			if v != "[not set]" {
				info.MemoryBytes, _ = strconv.ParseUint(v, 10, 64)
			}
		}
	}
	if m.Mode == User && info.Status == "active" && info.PID > 0 {
		info.Ports = getListeningPorts(info.PID)
	}
	return info, nil
}
func (m *Manager) GetStatusText(name string) (string, error) {
	if !m.ServiceExists(name) {
		return "", fmt.Errorf("service not found: %s", name)
	}
	out, err := m.exec().Output(m.Mode, "status", UnitName(name), "--no-pager", "-l")
	if err != nil {
		return err.Error(), nil
	}
	return out, nil
}
func (m *Manager) IsActive(name string) bool {
	return m.exec().StatusOK(m.Mode, "is-active", UnitName(name))
}
func parseSystemdTimestamp(ts string) (time.Time, bool) {
	ts = strings.TrimSpace(strings.TrimSuffix(ts, " UTC"))
	for _, layout := range []string{"Mon 2006-01-02 15:04:05", "Mon Jan 2 15:04:05 MST 2006", time.RFC3339} {
		if t, e := time.ParseInLocation(layout, ts, time.UTC); e == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

var pidPattern = regexp.MustCompile(`pid=(\d+)`)

func getListeningPorts(pid int) []int {
	if pid <= 0 {
		return nil
	}
	var ports []int
	seen := map[int]bool{}
	for _, args := range [][]string{{"-H", "-lntp"}, {"-H", "-lnup"}} {
		out, e := exec.Command("ss", args...).Output()
		if e != nil {
			continue
		}
		appendPortsForPID(&ports, seen, string(out), strconv.Itoa(pid))
	}
	sort.Ints(ports)
	return ports
}
func appendPortsForPID(ports *[]int, seen map[int]bool, out, pid string) {
	for _, line := range strings.Split(out, "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		owned := false
		for _, match := range pidPattern.FindAllStringSubmatch(line, -1) {
			if match[1] == pid {
				owned = true
				break
			}
		}
		if !owned {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) < 4 {
			continue
		}
		port := extractPort(fields[3])
		if port > 0 && !seen[port] {
			seen[port] = true
			*ports = append(*ports, port)
		}
	}
}
func extractPort(addr string) int {
	i := strings.LastIndex(addr, "]:")
	if i >= 0 {
		addr = addr[i+2:]
	} else if i = strings.LastIndex(addr, ":"); i >= 0 {
		addr = addr[i+1:]
	} else {
		return 0
	}
	n, _ := strconv.Atoi(addr)
	return n
}

func (m *Manager) ListTasks(all bool, serviceFilter string) ([]TaskInfo, error) {
	pattern := "smdctl-*-task-*.timer"
	if serviceFilter != "" {
		pattern = fmt.Sprintf("%s%s-task-*.timer", ServicePrefix, serviceFilter)
	}
	args := []string{"list-timers", "--output=json", "--no-pager", "--no-legend"}
	if all {
		args = append(args, "--all")
	}
	args = append(args, pattern)
	out, err := m.exec().Output(m.Mode, args...)
	if err != nil {
		return nil, err
	}
	var rows []map[string]json.RawMessage
	if err = json.Unmarshal([]byte(out), &rows); err != nil {
		return nil, err
	}
	result := make([]TaskInfo, 0)
	for _, row := range rows {
		var unit, activates string
		_ = json.Unmarshal(row["unit"], &unit)
		_ = json.Unmarshal(row["activates"], &activates)
		if !strings.HasPrefix(unit, ServicePrefix) || !strings.HasSuffix(unit, ".timer") {
			continue
		}
		base := strings.TrimSuffix(strings.TrimPrefix(unit, ServicePrefix), ".timer")
		parts := strings.SplitN(base, "-task-", 2)
		if len(parts) != 2 || (serviceFilter != "" && parts[0] != serviceFilter) {
			continue
		}
		result = append(result, TaskInfo{Service: parts[0], Task: parts[1], Mode: m.Mode, Next: formatTimerTimestamp(row["next"]), Last: formatTimerTimestamp(row["last"]), TimerUnit: unit, Activates: activates})
	}
	return result, nil
}
func formatTimerTimestamp(raw json.RawMessage) string {
	var n float64
	if len(raw) == 0 || json.Unmarshal(raw, &n) != nil || n == 0 {
		return "-"
	}
	return time.UnixMicro(int64(n)).Local().Format("2006-01-02 15:04")
}

func (m *Manager) GetServiceFile(name string) (string, error) {
	if !m.ServiceExists(name) {
		return "", fmt.Errorf("service not found: %s", name)
	}
	path, err := ServicePath(name, m.Mode)
	if err != nil {
		return "", err
	}
	b, err := os.ReadFile(path)
	return string(b), err
}
func DiscoverServiceMode(name string) (Mode, error) {
	for _, mode := range []Mode{User, System} {
		if p, err := ServicePath(name, mode); err == nil {
			if _, err = os.Stat(p); err == nil {
				return mode, nil
			}
		}
	}
	return "", fmt.Errorf("service not found: %s", name)
}
